### PR TITLE
Move SpiderDen file cache to use rwlock

### PR DESCRIPTION
**Pull request by revonateB0T**

### CODE REVIEW

File: app/src/main/kotlin/com/hippo/ehviewer/spider/SpiderDen.kt [hunk 1]
```diff
@@ -70,10 +73,11 @@ class SpiderDen(val info: GalleryInfo) {
     private val saveAsCbz = Settings.saveAsCbz
     private val archiveName = "$gid.cbz"
 
+    private val lock = ReentrantReadWriteLock()
```
Comment (FooIbar) [inline app/src/main/kotlin/com/hippo/ehviewer/spider/SpiderDen.kt:76]: Maybe we just use `ConcurrentHashMap`

Comment (revonateB0T) [reply app/src/main/kotlin/com/hippo/ehviewer/spider/SpiderDen.kt:76]: 之后我要换成MutableScatterMap的

File: app/src/main/kotlin/com/hippo/ehviewer/spider/SpiderDen.kt [hunk 2]
```diff
@@ -70,10 +73,11 @@ class SpiderDen(val info: GalleryInfo) {
...
+
     // Search in both directories to maintain compatibility
     private val fileCache by lazy {
-        (tempDownloadDir?.list().orEmpty() + downloadDir?.list().orEmpty())
-            .associateBy { it.name } as MutableMap
+        listOf(tempDownloadDir, downloadDir).mapNotNull { it?.list() }.flatten().associateBy { it.name } as MutableMap
```
Comment (FooIbar) [inline app/src/main/kotlin/com/hippo/ehviewer/spider/SpiderDen.kt:80]: nit: listOfNotNull

File: app/src/main/kotlin/com/hippo/ehviewer/spider/SpiderDen.kt [hunk 3]
```diff
@@ -162,9 +161,7 @@ class SpiderDen(val info: GalleryInfo) {
 
     private fun Path.findDownloadFileForIndex(index: Int, extension: String): Path {
         val name = perFilename(index, extension)
-        return synchronized(fileCache) {
-            fileCache.getOrPut(name) { resolve(name) }
-        }
+        return lock.write { fileCache.getOrPut(name) { resolve(name) } }
```
Comment (FooIbar) [inline app/src/main/kotlin/com/hippo/ehviewer/spider/SpiderDen.kt:164]: Using `getOrPut` won't leverage the benefits of RW lock.

File: app/src/main/kotlin/com/hippo/ehviewer/spider/SpiderDen.kt [hunk 4]
```diff
@@ -196,8 +193,9 @@ class SpiderDen(val info: GalleryInfo) {
             fops(tempFile)
             val file = resolve(tempFile.name.removeSuffix(TEMP_SUFFIX))
             file.delete()
+            lock.write { fileCache.remove(file.name) }
```
Comment (FooIbar) [inline app/src/main/kotlin/com/hippo/ehviewer/spider/SpiderDen.kt:196]: Why is this needed?

File: app/src/main/kotlin/com/hippo/ehviewer/spider/SpiderDen.kt (hunk 2 above)
Comment (revonateB0T) [reply app/src/main/kotlin/com/hippo/ehviewer/spider/SpiderDen.kt]: Done

**Review by FooIbar — APPROVED** [review]
